src/parameterFiles.ts

```diff
@@ -0,0 +1,507 @@
+  pickItems.sort((a, b) => {
+    const aData = a?.data;
+    const bData = a?.data;
+
+    // Close name matches go first
+    if (a?.data?.isCloseNameMatch !== b?.data?.isCloseNameMatch) {
+      return a?.data?.isCloseNameMatch ? -1 : 1;
+    }
+
+    // Otherwise compare filenames
+    // tslint:disable-next-line: strict-boolean-expressions
+    return (aData?.uri.fsPath || "").localeCompare(bData?.uri.fsPath || "");
+  });
+}
+
+function createQuickPickItem(paramFile: IPossibleParamFile, current: IPossibleParamFile | undefined, templateUri: Uri): IAzureQuickPickItem<IPossibleParamFile> {
+  const isCurrent: boolean = paramFile === current;
+  return {
+    label: `${isCurrent ? '$(check) ' : '$(json) '} ${paramFile.friendlyPath}`,
+    data: paramFile,
+    description: isCurrent ? currentMessage
+      : paramFile.isCloseNameMatch ? similarFilenameMessage
+        : undefined
+  };
+}
+
+function normalizePath(filePath: Uri | string): string {
+  const fsPath: string = typeof filePath === 'string' ? filePath :
+    filePath.fsPath;
+  let normalizedPath = path.normalize(fsPath);
+  if (isWin32) {
+    normalizedPath = normalizedPath.toLowerCase();
+  }
+
+  return normalizedPath;
+}
+
+/**
+ * Finds parameter files to suggest for a given template.
+ */
+export async function findSuggestedParameterFiles(templateUri: Uri): Promise<IPossibleParamFile[]> {
+  let paths: IPossibleParamFile[] = [];
+
+  // Current logic is simple: Find all .json/c files in the same folder as the template file and check
+  //   if they're a parameter file
+  try {
+    const folder = path.dirname(templateUri.fsPath);
+    const fileNames: string[] = await fse.readdir(folder);
+    for (let paramFileName of fileNames) {
+      const fullPath: string = path.join(folder, paramFileName);
+      const uri: Uri = Uri.file(fullPath);
+      if (await isParameterFile(fullPath)) {
+        paths.push({
+          uri,
+          friendlyPath: getFriendlyPathToParamFile(templateUri, uri),
+          isCloseNameMatch: mayBeMatchingParamFile(templateUri.fsPath, fullPath)
+        });
+      }
+    }
+  } catch (error) {
+    // Ignore
+  }
+
+  return paths;
+}
+
+async function isParameterFile(filePath: string): Promise<boolean> {
+  try {
+    if (!hasSupportedParamFileExtension(filePath)) {
+      return false;
+    }
+
+    if (await doesFileContainString(filePath, containsParamsSchema, readAtMostBytesToFindParamsSchema)) {
+      return true;
+    }
+  } catch (error) {
+    // Ignore
+  }
+
+  return false;
+}
+
+async function doesFileContainString(filePath: string, matches: (fileSubcontents: string) => boolean, maxBytesToRead: number): Promise<boolean> {
+  // tslint:disable-next-line: typedef
+  return new Promise<boolean>((resolve, reject) => {
+    const stream = fse.createReadStream(filePath, { encoding: 'utf8' });
+
+    let content: string = '';
+    stream.on('data', (chunk: string) => {
+      content += chunk;
+      if (matches(content)) {
```

Comment:
>matches [](start = 10, length = 7)

Nit: probably could look just in the chunk + length of desired match before that rather than looking in the entire content on every chunk.

Reply:
True, but can't just check the last chunk, it might be split between chunks. So we'd have to have knowledge about the length of the string to check, which bleeds implementation details from schema.ts.  I think probably not worth worrying about.